pkg/git/strategy/strategy_proxy_test.go

```diff
@@ -0,0 +1,255 @@
+					return goproxy.OkConnect, host
+				}
+				proxy.OnRequest().HandleConnect(proxyHandler)
+
+				return &git.AuthOptions{
+						Transport: git.HTTPS,
+						Username:  username,
+						Password:  password,
+						CAFile:    exampleCA,
+					}, func() {
+						os.RemoveAll(gitServer.Root())
+						gitServer.StopHTTP()
+					}
+			},
+			usedProxy: true,
+			getError:  false,
+		},
+		{
+			name:    "gogit_HTTP_PROXY",
+			gitImpl: gogit.Implementation,
+			url:     "http://example.com/bar/test-reponame",
+			branch:  "main",
+			setupGitProxy: func(g *WithT, proxy *goproxy.ProxyHttpServer, proxyGotRequest *bool) (*git.AuthOptions, func()) {
+				// Create the git server.
+				gitServer, err := gittestserver.NewTempGitServer()
+				g.Expect(err).ToNot(HaveOccurred())
+
+				username := "test-user"
+				password := "test-password"
+				gitServer.Auth(username, password)
+				gitServer.KeyDir(gitServer.Root())
+
+				g.Expect(gitServer.StartHTTP()).ToNot(HaveOccurred())
+
+				// Initialize a git repo.
+				err = gitServer.InitRepo("testdata/repo1", "main", "bar/test-reponame")
+				g.Expect(err).ToNot(HaveOccurred())
+
+				u, err := url.Parse(gitServer.HTTPAddress())
+				g.Expect(err).ToNot(HaveOccurred())
+
+				var proxyHandler goproxy.FuncReqHandler = func(req *http.Request, ctx *goproxy.ProxyCtx) (*http.Request, *http.Response) {
+					userAgent := req.Header.Get("User-Agent")
+					if strings.Contains(req.Host, "example.com") && strings.Contains(userAgent, "git") {
+						*proxyGotRequest = true
+						req.Host = u.Host
+						req.URL.Host = req.Host
+					}
+					return req, nil
+				}
+				proxy.OnRequest().Do(proxyHandler)
+
+				return &git.AuthOptions{
+						Transport: git.HTTP,
+						Username:  username,
+						Password:  password,
+					}, func() {
+						os.RemoveAll(gitServer.Root())
+						gitServer.StopHTTP()
+					}
+			},
+			usedProxy: true,
+			getError:  false,
+		},
+		{
+			name:    "gogit_HTTPS_PROXY",
+			gitImpl: gogit.Implementation,
+			url:     "https://github.com/git-fixtures/basic",
+			branch:  "master",
+			setupGitProxy: func(g *WithT, proxy *goproxy.ProxyHttpServer, proxyGotRequest *bool) (*git.AuthOptions, func()) {
+				var proxyHandler goproxy.FuncHttpsHandler = func(host string, ctx *goproxy.ProxyCtx) (*goproxy.ConnectAction, string) {
+					userAgent := ctx.Req.Header.Get("User-Agent")
+					if strings.Contains(host, "github.com") && strings.Contains(userAgent, "Go-http-client") {
+						*proxyGotRequest = true
+					}
+					return goproxy.OkConnect, host
+				}
+				proxy.OnRequest().HandleConnect(proxyHandler)
+
+				// go-git does not allow to use an HTTPS proxy and a custom root CA at the same time.
+				// See https://github.com/fluxcd/source-controller/pull/524#issuecomment-1006673163.
+				return nil, func() {}
+			},
+			usedProxy: true,
+			getError:  false,
+		},
+		{
+			name:    "gogit_NO_PROXY",
+			gitImpl: gogit.Implementation,
+			url:     "https://somewhere.com/bar/test-reponame",
```

Comment:
I'm not sure it is good practice to use `somewhere.com` which is a registered domain.
`example.org` would be a better choice, but even then I would personally opt for a [dummy IPv4 address](https://www.rfc-editor.org/rfc/rfc5737#section-3).
The current tests which use `example.com` don't ever reach that domain (unless the proxy is not used which would mean the test fails).
If the owner of `somewhere.com` sets up a Git repository at that specific URL with malicious intent, they could download arbitrary files on every tester's machine, which I wouldn't want to happen on my computer. Plus it would make the test fail wrongfully.

Also, it could be nice to test a `NO_PROXY` pattern as they are a big part of that variable.
So my suggestions would be:
```go
url:     "http://192.0.2.1/bar/test-reponame",
// ...
os.Setenv("NO_PROXY", "*.0.2.1")
```
Tell me what you think.

Thanks for the credits as co-author by the way. Metadata looks good to me.

Reply:
Good catch, I've updated as you suggested.